File: SafetyProgram.Core.Models.Serialization/XML/DocumentObjectXml.cs
using System;
using System.Xml.Linq;
using SafetyProgram.Core.IO;

namespace SafetyProgram.Core.Models.Serialization
{
    public sealed class DocumentObjectXml :
        IStorageConverter<IDocumentObject, XElement>
    {
        public XElement Store(IDocumentObject data)
        {
            throw new NotImplementedException();
        }

        public IDocumentObject Load(XElement data)
        {
            switch (data.Name.LocalName)
            {
                case "chemicaltable":
                    var chemicalSerializer = new ChemicalTableXml();
                    return chemicalSerializer.Load(data);
                default:
                    throw new NotImplementedException();
            }
        }
    }
}

File: SafetyProgram.Core.Models.Serialization/DocumentObjectXml.fs
module SafetyProgram.Core.Models.Serialization.DocumentObjectXml

open System.Xml.Linq
open ChemicalTableXml
open System
open SafetyProgram.Core.Models
open SafetyProgram.Base.Helpers
open FSharpx.Choice

let DocumentObjectXmlConverter : TwoWayConverter<DocumentObject, XElement> = {
    ConvertTo = fun _ -> new NotImplementedException() |> raise
    ConvertFrom = fun data -> 
        match data.Name.LocalName with
        | "chemicaltable" -> choose {
                let! x = ChemicalTableXmlConverter.ConvertFrom data
                return ChemicalTable x
            }         
        | _ -> Choice2Of2 "Unknown document object found."
}